### PR TITLE
build: replace pika with esbuild and tsc

### PR DESCRIPTION
**Pull request by wolfy1339**
Pika has been deprecated for a while now, and the project has now been archived.

Uses esbuild to transpile the TS source code into an ESM source, NodeJs bundle, and a browser bundle

Uses the TypeScript compiler to generate the types

----

### Before the change?

* Uses pika for the build system

### After the change?

* Uses esbuild to generate the JS bundles
* Uses the TypeScript compiler to generate the type definitions

### Other information

*

----

### Pull request checklist
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Added the appropriate label for the given change

### Does this introduce a breaking change?

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/main/community/breaking_changes.md) to help!

- [ ] Yes (Please add the `Type: Breaking change` label)
- [x] No

If `Yes`, what's the impact:

* N/A

### Pull request type

Please add the corresponding label for change this PR introduces:
- Dependencies/code cleanup: `Type: Maintenance`

----

### CODE REVIEW

**Comment by kfcampbell** [comment]
I'm :+1: on esbuild and tsc, and this approach seems reasonable to me.

File: scripts/build.ts
Comment (oscard0m) [inline scripts/build.ts]: We are missing the format of 
```suggestion
  copyFileSync("LICENSE.md", "pkg/LICENSE.md");
```

**Review by oscard0m — COMMENTED** [review]
# ❤️‍🔥 Feedback
Amazing job @wolfy1339! Thanks for taking the time and effort on this. This will have a lot of benefits for the community and us 👏🏽 🥳 

I checked your changes and ran `diff` to compare both outputs. I see some differences, but I'm unsure if they are expected.

`Node version used`: `14.21.3`
`npm version used`: `9.6.0`

## 📦⚖️  Output differences
I see some differences which I'm not sure if they are expected or not:
### `pkg/LICENSE.md`
missing `.md` extension. Is it intended?

### `pkg/package.json`
- The `author` appears with ESBuild. I think this is an improvement 👏🏽 
- Fixed `files` patterns ✅ 
- With `pika`, the `main`, `module`, `types` and `source` paths were not starting with a relative path (`./`). Is this an intended change?

### `pkg/dist-node/index.js` and `pkg/dist-web/index.js`
I see noticeable file size differences:
image

image

Checking the differences, I see some variable naming differences and spacing/indentation, so I'm assuming ESBuild does a better job there. Still, the file size increases ⚠️ from `19kb` to `21kb`, which I guess is something we don't want.

### `pkg/dist-src/*`
We are emitting source maps, but I think we don't want the `dist-src/`, right? I guess the non-source-map files increased their size because of sourcemaps.
image

### `pkg/dist-types`
It looks identical to me 🍾 ✅ 


## 📊 Benchmarks
I leave here some screenshots of benchmarks I run on my local machine.
**command:** `hyperfine --runs 3 'npm run build'`

I don't have much experience with `esbuild` but I was expecting better execution times. Maybe is `tsc` consuming our time here? Or maybe the `js` file invoking `esbuild`? Since the time is almost the same this is something I would be happy to nerd about in a different PR 🤓 unless you see any easy quick-win here.

I would be happy to support you here. Let me know if you want to split the work or 
### Pika (`main` branch)
image

### ESBuild (`esbuild` branch)
image

## Migration Plan
Do you have any plan in mind for extending this to the rest of the repositories of the ecosystem? I'm happy to help with octoherd-scripts or any other approach we plan to do to migrate this.

File: package.json
Comment (oscard0m) [inline package.json:55]: We need to update `.gitignore` comment as well: `// # Pika builds`

File: scripts/build.ts
Comment (oscard0m) [inline scripts/build.ts]: Would it speed things up if we wrap these commands in a `Promise.all()`?
```js
await Promise.all([
  esbuild.build(...), // dist-node
  esbuild.build(...), // dist-web
  esbuild.build(...), // dist-src
])`
```

File: scripts/build.ts
Comment (oscard0m) [inline scripts/build.ts]: 💅🏽 It would be nice to correlate this with `packageJson.engines.node`, but we can tackle this in a different PR.

```suggestion
    target: pkgJson.engines.node,
```

File: scripts/build.ts
Comment (oscard0m) [inline scripts/build.ts]: I would check if making the `read/write` actions async gives us some benefits.

**Comment by wolfy1339** [comment]
> With pika, the main, module, types and source paths were not starting with a relative path (./). Is this an intended change?

I can fix that

> I don't have much experience with esbuild but I was expecting better execution times. Maybe is tsc consuming our time here?

I'm not sure. We can certainly play around with this later.
One thing that might help is not using `ts-node` for the build script.

> Do you have any plan in mind for extending this to the rest of the repositories of the ecosystem? I'm happy to help with octoherd-scripts or any other approach we plan to do to migrate this.

I didn't have any specific plan.
An octoherd script can do most of the work, but since we modify `package.json` and `package-lock.json` we will need manual input

File: scripts/build.ts
Comment (wolfy1339) [reply scripts/build.ts]: We have to do `build-src` first, as the other builds are based off of it.

For the other 2, we can wrap them in `Promise.all()`

File: scripts/build.ts
Comment (oscard0m) [inline scripts/build.ts]: 💅🏽 I found this plugin: https://github.com/AkaraChen/esbuild-plugin-clean/blob/main/esm.mjs. 

Maybe we can use it or take it as inspiration for our purposes: `build.onStart(...)`

File: scripts/build.ts
Comment (oscard0m) [inline scripts/build.ts]: I found these 2 plugins which maybe we want to keep track of:
- https://github.com/nickjj/esbuild-copy-static-files
- https://github.com/sprout2000/esbuild-copy-plugin
- https://github.com/ozanozbek/esbuild-plugin-copy-file

File: package.json
Comment (oscard0m) [inline package.json]: Maybe we can get better timings with [`tsup`](https://tsup.egoist.dev/#generate-declaration-file) by esbuild? I will experiment with that in the upcoming days but I don't want to block you.

File: scripts/build.ts
Comment (wolfy1339) [reply scripts/build.ts]: It would be nice, but the `engines` field isn't just a node version, it's a semver expression

Comment (oscard0m) [reply scripts/build.ts]: True. It's not an easy task.

File: scripts/build.ts
Comment (wolfy1339) [reply scripts/build.ts]: Done in 1cdd9b4f15c2d85ad5d54a76006e234a42addf94

File: scripts/build.ts
Comment (wolfy1339) [reply scripts/build.ts]: Switching to the Promise based `fs` api, and using async glob, it is roughly a second faster

File: package.json
Comment (wolfy1339) [reply package.json]: We can try `tsup`, I'm not sure if you can specify different output folders for the different builds.

I just went this way to keep the same package structure

File: package.json
Comment (wolfy1339) [reply package.json:55]: Done in 7d570d736f90104e09ed029ccb95a6d84de72766

**Comment by wolfy1339** [comment]
With all changes since your review @oscard0m, the build time has reduced by 2 seconds.
```
$ hyperfine --runs 3 'npm run build'
Benchmark 1: npm run build
  Time (mean ± σ):      7.026 s ±  0.204 s    [User: 14.463 s, System: 0.413 s]
  Range (min … max):    6.864 s …  7.255 s    3 runs
```

That's 4 seconds faster than Pika

```
$ node --version
v18.14.0
$ npm --version
9.3.1
```

**Comment by wolfy1339** [comment]
Should we change the current value of the `module` field to point to `dist-src/` and add a new `browser` field that points to `dist-web/` ?

This would alleviate issues like https://github.com/octokit/auth-app.js/issues/59 until work on `octokit-next` is finished

```diff
diff --git a/scripts/build.mjs b/scripts/build.mjs
index 3ef8688..73d9ef9 100644
--- a/scripts/build.mjs
+++ b/scripts/build.mjs
@@ -72,9 +71,9 @@ async function main() {
         ...pkg,
         files: ["dist-*/**", "bin/**"],
         main: "dist-node/index.js",
-        module: "dist-web/index.js",
+        browser: "dist-web/index.js",
         types: "dist-src/index.d.ts",
-        source: "dist-src/index.js",
+        module: "dist-src/index.js",
       },
```

**Comment by ollie-iterators** [comment]
What is blocking this PR from getting merged?

**Comment by wolfy1339** [comment]
Nothing specific at the moment 

It just needs a review and approval from another maintainer

**Comment by oscard0m** [comment]
@ollie-iterators @wolfy1339 I plan to review this on Thursday. 

Sorry for the late reply here.

File: scripts/build.mjs
Comment (oscard0m) [inline scripts/build.mjs]: This line fails in local development when the `pkg/` folder does not exist. Maybe we can apply the [`force: true` option](https://nodejs.org/docs/latest-v18.x/api/fs.html#fsrmpath-options-callback) here?

```suggestion
  await rm("pkg", { recursive: true, force: true });
```

**Comment by oscard0m** [comment]
> Should we change the current value of the `module` field to point to `dist-src/` and add a new `browser` field that points to `dist-web/` ?
> 
> This would alleviate issues like [octokit/auth-app.js#59](https://github.com/octokit/auth-app.js/issues/59) until work on `octokit-next` is finished
> 
> ```diff
> diff --git a/scripts/build.mjs b/scripts/build.mjs
> index 3ef8688..73d9ef9 100644
> --- a/scripts/build.mjs
> +++ b/scripts/build.mjs
> @@ -72,9 +71,9 @@ async function main() {
>          ...pkg,
>          files: ["dist-*/**", "bin/**"],
>          main: "dist-node/index.js",
> -        module: "dist-web/index.js",
> +        browser: "dist-web/index.js",
>          types: "dist-src/index.d.ts",
> -        source: "dist-src/index.js",
> +        module: "dist-src/index.js",
>        },
> ```

I would do that in a separate PR (feel free to create a new issue mentioning this comment) so we do not add more things on top of this migration. I think it's good the differences are minimum in this PR. What do you think?

File: scripts/build.mjs
Comment (oscard0m) [inline scripts/build.mjs]: I'm seeing noticeable differences in the generated sourcemaps from one version and the other. The 2 main differences I see are: 

- `Pika`'s generated sourcemaps are pointing to `dist-src/` and the ones generated by ESBuild are pointing to `src/`. I guess we don't want this so we need the `sourcemaps: "external"` option. Does this make sense?
- `Pika`'s generated sourcemaps files are not including `sourcesContent` but the ones generated by ESBuild yes. I guess we don't want to emit that by using `sourcesContent: false`. What do you think?

```suggestion
  sourcemap: external,
  sourcesContent: false
```

File: scripts/build.mjs
Comment (oscard0m) [inline scripts/build.mjs:23]: We were not emitting sourcemaps in `dist-src/` before so I think we should preserve that.

```suggestion
    ...sharedOptions,
    sorucemaps: false
```

File: scripts/build.mjs
Comment (oscard0m) [inline scripts/build.mjs]: comparing outputs, seems like `pika` was generating the `package.json` file with [`sideEffects: false`](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free). To prevent this to be potentially a breaking change, I would keep it.

```suggestion
      {
        ...pkg,
        files: ["dist-*/**", "bin/**"],
        main: "dist-node/index.js",
        module: "dist-web/index.js",
        types: "dist-src/index.d.ts",
        source: "dist-src/index.js",
        sideEffects: false
      },
```

**Review by oscard0m — CHANGES_REQUESTED** [review]

File: scripts/build.mjs
Comment (wolfy1339) [reply scripts/build.mjs:23]: I'll have to test that to make sure it doesn't negatively effect the other builds

File: scripts/build.mjs
Comment (wolfy1339) [reply scripts/build.mjs]: Wouldn't it be better to have them point to the source TS files instead of `dist-src/`?

That way the files can stay independent from each other

**Comment by wolfy1339** [comment]
> > Should we change the current value of the `module` field to point to `dist-src/` and add a new `browser` field that points to `dist-web/` ?
> > 
> > This would alleviate issues like [octokit/auth-app.js#59](https://github.com/octokit/auth-app.js/issues/59) until work on `octokit-next` is finished
> > 
> > ```diff
> > diff --git a/scripts/build.mjs b/scripts/build.mjs
> > index 3ef8688..73d9ef9 100644
> > --- a/scripts/build.mjs
> > +++ b/scripts/build.mjs
> > @@ -72,9 +71,9 @@ async function main() {
> >          ...pkg,
> >          files: ["dist-*/**", "bin/**"],
> >          main: "dist-node/index.js",
> > -        module: "dist-web/index.js",
> > +        browser: "dist-web/index.js",
> >          types: "dist-src/index.d.ts",
> > -        source: "dist-src/index.js",
> > +        module: "dist-src/index.js",
> >        },
> > ```
> 
> I would do that in a separate PR (feel free to create a new issue mentioning this comment) so we do not add more things on top of this migration. I think it's good the differences are minimum in this PR. What do you think?

Sure thing 👍

Let's do it that way

File: scripts/build.mjs
Comment (oscard0m) [reply scripts/build.mjs]: We are not publishing the `/src/` folder with the `.ts` files so I think that would not work. Am I correct in my assumption?

image

File: package.json
Comment (oscard0m) [reply package.json]: Sure, let's not add more complexity to this and stay with what we have.

File: scripts/build.ts
Comment (oscard0m) [reply scripts/build.ts]: Let's keep this for a future and not add more complexity to this PR

File: scripts/build.ts
Comment (oscard0m) [reply scripts/build.ts]: Let's keep this for a future and not add more complexity to this PR

**Comment by oscard0m** [comment]
> I didn't have any specific plan. An octoherd script can do most of the work, but since we modify `package.json` and `package-lock.json` we will need manual input

Let's have the first version for `webhooks.js` and once we are happy we work an octoherd script together :)

**Comment by oscard0m** [comment]
@wolfy1339 let me know if I can help you with this in some way. I have some availability for pair-program as well.

File: scripts/build.mjs
Comment (wolfy1339) [reply scripts/build.mjs:23]: This makes it so that the other builds reference `dist-src/`

File: scripts/build.mjs
Comment (wolfy1339) [reply scripts/build.mjs]: The map contains the source files within the `sourceContent` property. Which makes them completely indepedent

File: scripts/build.mjs
Comment (wolfy1339) [reply scripts/build.mjs:23]: Let's just make it output the same as Pika for now, and we can revisit these points later

File: scripts/build.mjs
Comment (wolfy1339) [reply scripts/build.mjs]: Let's just make it output the same as Pika for now, and we can revisit these points later

**Comment by wolfy1339** [comment]
I addressed the review comments and rebased the branch

**Review by oscard0m — DISMISSED** [review]
Awesome job @wolfy1339 ! Thanks for the time and effort. Do you think it's interesting to do a beta release first? I would like to try upgrading the package in local for other octokit repos and also try the sourcemaps/bundle size once published in npm.

**Comment by wolfy1339** [comment]
We could definitely do a beta release first

**Comment by wolfy1339** [comment]
Actually, that would conflict with the current `beta` branch which holds v11 in #792

**Comment by wolfy1339** [comment]
This won't trigger a release, as it only touches the build system

**Comment by oscard0m** [comment]
> This won't trigger a release, as it only touches the build system

Or we mark it as a breaking change (in alpha) or we do a second PR in alpha which triggers a release.

Going to bed now. I read you tomorrow morning CET

**Comment by wolfy1339** [comment]
> > This won't trigger a release, as it only touches the build system
> 
> Or we mark it as a breaking change (in alpha) or we do a second PR in alpha which triggers a release.
> 
> Going to bed now. I read you tomorrow morning CET

The semantic release configuration doesn't have alpha set up.

Anyways, I mirrored this branch to `next` and created an empty commit to trigger a patch release. It is available on the `@next` npm tag

File: scripts/build.mjs
Comment (oscard0m) [inline scripts/build.mjs]: I'm trying [this branch](https://github.com/octokit/app.js/pull/405) for `@octokit/app.js` and found the following issue:

```suggestion
        types: "dist-types/index.d.ts",
```

File: scripts/build.mjs
Comment (oscard0m) [inline scripts/build.mjs]:
```suggestion
  sourcesContent: true,
```
> Let's just make it output the same as Pika for now, and we can revisit these points later

I'm sorry here, I don't know what I was checking but the `sourceContent` is served in Pika so I think we should be keeping it for ESBuild for the moment.

**Comment by oscard0m** [comment]
Thanks @wolfy1339! Do you mind pinging me here when `next` branch is up to date with these changes and release in @next done so I try again the test I'm doing in `@octokit/app.js`?

If you don't have time because is late in your timezone I can take it.

---

EDIT: I'm applying changes to `next` branch FYI

**Comment by oscard0m** [comment]
The changes look good to me. I applied them to `next` branch ([`v10.7.3`](https://www.npmjs.com/package/@octokit/webhooks/v/10.7.3)) and [the draft PR of `@octokit/app.js`](https://github.com/octokit/app.js/pull/405) using this version looks good.

There are some small differences but I'm assuming these are differences/optimizations ESBuild does that Pika was doing in a different way.

**Review by oscard0m — DISMISSED** [review]

**Comment by oscard0m** [comment]
@wolfy1339 I had an idea on what we could do to make it easier the scaffolding for the rest of the repositories (and maintain it in an easy way): https://github.com/octokit/webhooks.js/pull/829

Let me know your thoughts here.

**Review by oscard0m — APPROVED** [review]

**Comment by oscard0m** [comment]
@wolfy1339 I resolved the conflicts. Are we ready to merge? To trigger a release, should the PR title be `fix` instead?

**Comment by wolfy1339** [comment]
The releases are kinda screwed up.

Since the `next` branch has a few releases that take up some patch versions, we'll need to merge that into `main` for it to trigger a proper release.

It's not ideal, as we didn't have alpha releases enabled and the beta release was taken up by `v11`

**Comment by oscard0m** [comment]
> The releases are kinda screwed up.
> 
> Since the `next` branch has a few releases that take up some patch versions, we'll need to merge that into `main` for it to trigger a proper release.
> 
> It's not ideal, as we didn't have alpha releases enabled and the beta release was taken up by `v11`

Is not possible to have 10.7.1 in `latest` even we are in 10.7.3 in `next`?

**Comment by wolfy1339** [comment]
Nope. You can only use a version number once, no matter the release channel.

You can overwrite a release, but I don't think semantic release is capable

**Comment by oscard0m** [comment]
> Nope. You can only use a version number once, no matter the release channel.
> 
> You can overwrite a release, but I don't think semantic release is capable

Let's merge `next` in `main` then. The release notes look decent to me.

**Pull request closed without merging**
